src/Model.integration.test.ts

```diff
@@ -0,0 +1,279 @@
+        ably,
+        channelName,
+        logger,
+        syncOptions: defaultSyncOptions,
+        optimisticEventOptions: defaultOptimisticEventOptions,
+        eventBufferOptions: defaultEventBufferOptions,
+      },
+    );
+
+    context.model = model;
+    context.channelName = channelName;
+    context.ably = ably;
+
+    context.eventData = {
+      name: 'update',
+      data: {
+        foo: 34,
+      },
+      mutationId: 'some-id-1',
+    };
+    context.syncData = syncData;
+  });
+
+  afterEach<TestContext>(async ({ model }) => {
+    await model.dispose();
+    vi.restoreAllMocks();
+  });
+
+  it<TestContext>('changes state on sync, pause, resume, dispose', async ({ model }) => {
+    expect(model.state).toEqual('initialized');
+
+    const sync = model.sync(1);
+    expect(model.state).toEqual('syncing');
+
+    await sync;
+    expect(model.state).toEqual('ready');
+
+    await model.pause();
+    expect(model.state).toEqual('paused');
+
+    await model.resume();
+    expect(model.state).toEqual('ready');
+
+    await model.dispose();
+    expect(model.state).toEqual('disposed');
+  });
+
+  it<TestContext>('successfully sets the data from the event in optimistic()', async ({
+    ably,
+    model,
+    channelName,
+    eventData,
+    syncData,
+  }) => {
+    await model.sync(1);
+    expect(model.data.optimistic).toEqual(syncData[1].data);
+
+    let subscription = new Subject<void>();
+    const subscriptionCalls = getEventPromises(subscription, 3);
+    const subscriptionSpy = vi.fn(() => subscription.next());
+    const finalData = { ...syncData[1].data, ...eventData.data };
+
+    model.subscribe(subscriptionSpy);
+
+    await subscriptionCalls[0];
+    expect(model.data.confirmed).toEqual(syncData[1].data);
+
+    const [confirmation] = await model.optimistic(eventData);
+
+    await subscriptionCalls[1];
+    expect(model.data.optimistic).toEqual(finalData);
+
+    const channel = ably.channels.get(channelName);
+    await channel.publish({
+      data: eventData.data,
+      name: 'update',
+      extras: {
+        headers: {
+          'x-ably-models-event-uuid': eventData.mutationId,
+        },
+      },
+    });
+
+    await subscriptionCalls[2];
+    await confirmation;
+
+    expect(subscriptionSpy).toHaveBeenCalledTimes(3);
+    expect(subscriptionSpy).toHaveBeenNthCalledWith(3, null, finalData);
+    expect(model.data.confirmed).toEqual(finalData);
+  });
+
+  it<TestContext>('rejects the data in optimistic() and rolls back the changes if back-end rejects the data', async ({
+    ably,
+    model,
+    channelName,
+    eventData,
+    syncData,
+  }) => {
+    await model.sync(1);
+    let subscription = new Subject<void>();
+    const subscriptionCalls = getEventPromises(subscription, 3);
+    const subscriptionSpy = vi.fn(() => subscription.next());
+    const finalData = { ...syncData[1].data, ...eventData.data };
+
+    model.subscribe(subscriptionSpy);
+    await model.optimistic(eventData);
```

Comment:
last missing piece: let's capture the confirmation promise here:

```
const [confirmation] = await model.optimistic(eventData);
```

And then later, assert it rejects:

```
await expect(confirmation).rejects.toThrow()
```